README.md
# HSVColorController, BgColorControler, NgColorController
* new NgColorController , fixed style of color-controller (no popup, fixed height:125px, huebar width from 15px to 20px)
* new BgColorController for handling two colors (i.e. foreground-background color)
* new HSVColorController (toggle Mode: RGB <-> HSV)

![BgColorController](screen5.jpg)
![NgColorController](screen6.jpg)
![HSVColorController-HSV](screen3.jpg)
![HSVColorController-RGB](screen4.jpg)

### Usage
* addNgColor(properties,property) for NgColorController
* addBgColor(properties,property) for BgColorController
* addHSVColor(properties,property) for HSVColorController

Note: 
* For using BgColorController displaying a two colors, a second "property+bg" must be present. For Example: 
`var myProps={color:#ffffff,colorbg:#000000};
datgui.addBgColor(myProps,color);`
* Changing the height for Ng/BgColorControllers can be done by adding the height field in the CSS .dg.c tag: `
 .c {
    float: left;
    width: 60%;
    height:125px; /*JLA*/
    position: relative;
  }
`. The height of HSVColorController is fixed internaly to 125px.

# UI Controller (dat.gui ported to commonjs)

## dynamic_properties branch

This branch add the possibility for controllers to handle "dynamic properties".

It means that the property supplied to the controller may not exist as such on the supplied object but in the form of a couple of standard getter/setter functions that match the property's name.

Ie: the property `"position"` match the `getPosition()`/`setPosition()` functions.

---
Original dat.GUI readme below:

---

# dat.GUI

A lightweight graphical user interface for changing variables in JavaScript.

Get started with dat.GUI by reading the [tutorial](http://workshop.chromeexperiments.com/examples/gui)
or the [API documentation](API.md).

The dat.GUI library needs to be built with the JS and CSS separately. 
This can be turned on and off by changing one line in `rollup.config.js`:
```js
plugins: [
    ...
    sass({
        ...
        insert: false
        ...
    })
    ...
]
```

This is because, when `insert` is set to `true`, the built dat.GUI file 
appends the CSS file to the HTML, which means you can't control the order 
that it is included. Because any website needs to be able to override 
the default styles, the default styles need to be included before the overrides. 



## This fork's extra features

* Added hide/show methods to all Controllers ([#93](https://github.com/dataarts/dat.gui/issues/93)).
* Delete button added to preset menu ([#215](https://github.com/dataarts/dat.gui/issues/215)).
* Pass passive option to most event listeners to improve scroll performance ([#193](https://github.com/dataarts/dat.gui/issues/193)).
* Added forceUpdateDisplay to listen(). Allows overiding of the active check on GUI inputs before updating ([#212](https://github.com/dataarts/dat.gui/issues/212)).
* Fixed left border color not working on non-hex color controllers ([#208](https://github.com/dataarts/dat.gui/issues/208))
* Added value plotter controller ([#191](https://github.com/dataarts/dat.gui/issues/191))
* Implemented number box changes on keyboard arrow up/down, mouse wheel scroll and converted inputs to number. ([#195](https://github.com/dataarts/dat.gui/issues/195))

This is a fork of dat.gui where I merged some outstanding pull request to fix bugs.
Also includes @MacroMan's fork and others' with extra features.

![dat.gui preview of all inputs](dat_panel.gif)

See [API.md](API.md) for usage


## Packaged Builds

The easiest way to use dat.GUI in your code is by using the built source at [`build/dat.gui.js`](build/dat.gui.js). This built JavaScript file bundles all the necessary dependencies to run dat.GUI.

In your `head` tag, include the following code:

```html
<script type="text/javascript" src="dat.gui.min.js"></script>
```


## Example code

See `tests/manual.html` for an example of each type of controller.


## Installing from npm

Install the module:

```bash
$ npm install --save dat.gui
```

Use it:

```js
// CommonJS:
const dat = require('dat.gui');

// ES6:
import * as dat from 'dat.gui';

var obj = { x: 5 };
var gui = new dat.GUI();

gui.add(obj, 'x').onChange(function() {
  // obj.x will now have updated value
});
```




## Directory Contents

+ `build` - Concatenated source code for browsers.
+ `src` - source code in commonjs format.
+ `tests` - [QUnit](https://github.com/jquery/qunit) test suite.
+ `util` - [node.js](http://nodejs.org/) utility scripts for compiling source.



## Building your own dat.GUI

In the terminal, enter the following:

```bash
$ npm install
$ npm run build
```

This will create a rolled-up build of dat.GUI at `build/dat.gui.js` and its
minified version at `build/dat.gui.min.js`.



## Running tests

In the terminal, enter `npm run dev` and open http://localhost:8080/tests/

* Open `qunit.html` to run automated unit tests
* Open `manual.html` for a manual test suite - events are logged to the browser console.


## npm scripts

* npm run dev - Build development version of script and watch for changes.
* npm run build - Build development and production version of scripts.
* npm run build-docs - Automatically generate API.md from source code
* npm run lint - Run eslint on all code


## Working with Content Security Policy

If you're using a server with a Content Security Policy in place that blocks 'unsafe-inline', you will have problems when dat.gui.js tries to inject style information. To get around this, load 'build/dat.gui.css' as an external style sheet.


## Changes

View the [Change Log](CHANGELOG.md)


## Thanks

The following libraries / open-source projects were used in the development of dat.GUI:

 * [Rollup](https://rollupjs.org)
 * [Sass](http://sass-lang.com/)
 * [Node.js](http://nodejs.org/)
 * [QUnit](https://github.com/jquery/qunit) / [jquery](http://jquery.com/)
 * [stats.js](http://mrdoob.github.io/stats.js/)



CHANGELOG.md
## Changelog

### 0.7.6

#### 1.2.0+dataarts-0.7.6
* Implemented number box changes on keyboard arrow up/down, mouse wheel scroll and converted inputs to number. ([#195](https://github.com/dataarts/dat.gui/issues/195))

#### 1.1.0+dataarts-0.7.6
* Added value plotter controller, allowing tracking of basic numberic values over time. ([#191](https://github.com/dataarts/dat.gui/issues/191))

#### 1.0.0+dataarts-0.7.6
* Updated version number to be compliant with Semantic Versioning 2.0.0
 * First part is this projects version, starting at 1.0.0 to signify that it is production ready
 * Second part (after the +) is the last merge point of the [upstream project](https://github.com/dataarts/dat.gui)
* Brought all npm packages up to date and resolved vulnerabilities

#### 0.7.5-0.1.1
* Fixed left border color not working on non-hex color controllers ([#208](https://github.com/dataarts/dat.gui/issues/208))

#### 0.7.5-0.1.0
* Added hide/show methods to all Controllers ([#93](https://github.com/dataarts/dat.gui/issues/93)).
* Delete button added to preset menu ([#215](https://github.com/dataarts/dat.gui/issues/215)).
* Pass passive option to most event listeners to improve scroll performance ([#193](https://github.com/dataarts/dat.gui/issues/193)).
* Added forceUpdateDisplay to listen(). Allows overiding of the active check on GUI inputs before updating ([#212](https://github.com/dataarts/dat.gui/issues/212)).


### 0.7.1
 * Fix listener cleanup in .destroy() and .removeFolder(). #177

### 0.7.0
 * Changed build system to Rollup, publishing ES6 and UMD modules. #172
 * Added .removeFolder(). #113 #158
 * Added API documentation. #165
 * Improved touch support. #173

### 0.6.5
 * Add browserify support. #137

### 0.6.4
 * Fixed formatting issue on Windows. #136
 * Fixed issue with color selector getting chopped off at bottom of gui. #136

### 0.6.3
 * Added ability to put close button on top of the menu [ex: var gui = new dat.gui.GUI({closeOnTop:true})]. #106 #122
 * Fixed issue with layout not updating on page resize.
 * Fixed color picker showing in wrong position on user scroll (long menu). #105
 * Fixed checkbox weirdness while listening. #40
 * Fixed bower issues. #125

### 0.6.2
 * Fixed issue with color picker not returning correct values onFinishChange. #116 #117
 * Fixed issue with color picker not showing the correct label for color format. #117
 * Fixed NumberControllerBox onFinishChange event. #112
 * Fixed issue with GUI not being sized correct at start. #111

### 0.6.1
 * Fixed issue with color picker not working on a page that has scrolled. #37
 * Fixed issue with sliders created with min()/max() not remembering their name or to listen. #107

### 0.6.0
 * Ported to ES6
 * Exported using Universal Module Definition (UMD) for max compatibility (Commonjs, Requirejs, AMD, global var)
 * Now using webpack for build
 * Optional external CSS file (dat.gui.css) for use on CSP-enabled servers that block unsafe-inline
 * Added updateDisplay() to GUI, to update all controls in all folders - https://github.com/dataarts/dat.gui/pull/97
 * Fixed GUI.destroy() to remove all window eventListeners - https://github.com/dataarts/dat.gui/pull/88
 * Fixed performance issue when rotated on tablet/mobile - https://github.com/dataarts/dat.gui/pull/91
 * Fixed issue that prevented user from changing values of controls that are listening - https://github.com/dataarts/dat.gui/issues/100
 * Fixed issues with onFinishChange callbacks on revert - https://github.com/dataarts/dat.gui/pull/103
 * Fixed issues with color selector formatting - https://github.com/dataarts/dat.gui/issues/73
 * Fixed issues with step parameters in sliders - https://github.com/dataarts/dat.gui/issues/74
 * Fixed an issue with colors based on arrays - https://github.com/dataarts/dat.gui/pull/57
 * Fixed factory.js, Step param was not used - https://github.com/dataarts/dat.gui/pull/45

### 0.5
 * Moved to requirejs for dependency management.
 * Changed global namespace from *DAT* to *dat* (lowercase).
 * Added support for color controllers. See [Color Controllers](http://workshop.chromeexperiments.com/examples/gui/#4--Color-Controllers).
 * Added support for folders. See [Folders](http://workshop.chromeexperiments.com/examples/gui/#3--Folders).
 * Added support for saving named presets.  See [Presets](http://workshop.chromeexperiments.com/examples/gui/examples/gui/#6--Presets).
 * Removed `height` parameter from GUI constructor. Scrollbar automatically induced when window is too short.
 * `dat.GUI.autoPlace` parameter removed. Use `new dat.GUI( { autoPlace: false } )`. See [Custom Placement](http://workshop.chromeexperiments.com/examples/gui/#9--Custom-Placement).
 * `gui.autoListen` and `gui.listenAll()` removed. See [Updating The Display Manually](http://workshop.chromeexperiments.com/examples/gui/#11--Updating-the-Display-Manually).
 * `dat.GUI.load` removed. See [Saving Values](http://workshop.chromeexperiments.com/examples/gui/#5--Saving-Values).
 * Made Controller code completely agnostic of GUI. Controllers can easily be created independent of a GUI panel.

### 0.4
 * Migrated from GitHub to Google Code.


demo/demo.js
function FizzyText(message, font) {

  var that = this;

  // These are the variables that we manipulate with gui-dat.
  // Notice they're all defined with "this". That makes them public.
  // Otherwise, gui-dat can't see them.


  this.growthSpeed = 0.2;       // how fast do particles change size?
  this.maxSize = 5.59;          // how big can they get?
  this.noiseStrength = 10;      // how turbulent is the flow?
  this.speed = 0.4;             // how fast do particles move?
  this.displayOutline = false;  // should we draw the message as a stroke?
  this.framesRendered = 0;
  this.font = 0;
  this.allFonts = ['arial', 'times', 'courier'];
	
  // __defineGetter__ and __defineSetter__ makes JavaScript believe that
  // we've defined a variable 'this.message'. This way, whenever we
  // change the message variable, we can call some more functions.

  this.__defineGetter__("message", function () {
    return message;
  });

  this.__defineSetter__("message", function (m) {
    message = m;
    createBitmap(message, font);
  });

  this.__defineGetter__("font", function () {
    return font;
  });

  this.__defineSetter__("font", function (f) {
    font = f;
    createBitmap(message, font);
  });

  // We can even add functions to the DAT.GUI! As long as they have
  // 0 arguments, we can call them from the dat-gui panel.

  this.explode = function() {
    var mag = Math.random() * 30 + 30;
    for (var i in particles) {
      var angle = Math.random() * Math.PI * 2;
      particles[i].vx = Math.cos(angle) * mag;
      particles[i].vy = Math.sin(angle) * mag;
    }
  };

  ////////////////////////////////////////////////////////////////

  var _this = this;

  var width = 550;
  var height = 200;
  var textAscent = 101;
  var textOffsetLeft = 80;
  var noiseScale = 300;
  var frameTime = 30;

  var colors = ["#00aeff", "#0fa954", "#54396e", "#e61d5f"];

  // This is the context we use to get a bitmap of text using
  // the getImageData function.
  var r = document.createElement('canvas');
  var s = r.getContext('2d');

  // This is the context we actually use to draw.
  var c = document.createElement('canvas');
  var g = c.getContext('2d');

  r.setAttribute('width', width);
  c.setAttribute('width', width);
  r.setAttribute('height', height);
  c.setAttribute('height', height);

/*
	var onResize = function() {
		r.width = c.width = width = window.innerWidth;
		r.height = c.height = height = window.innerHeight;
		console.log(width, height);
	}
	
	window.addEventListener('resize', function() {
		onResize();
		createBitmap(this.message);
	}, false);
	onResize();
*/

  // Add our demo to the HTML
  document.getElementById('helvetica-demo').appendChild(c);

  // Stores bitmap image
  var pixels = [];

  // Stores a list of particles
  var particles = [];

  // Set g.font to the same font as the bitmap canvas, incase we
  // want to draw some outlines.
  s.font = g.font = "800 82px " + this.allFonts[ this.font ];
  
  // Instantiate some particles
  for (var i = 0; i < 1000; i++) {
    particles.push(new Particle(Math.random() * width, Math.random() * height));
  }

  // This function creates a bitmap of pixels based on your message
  // It's called every time we change the message property.
  var createBitmap = function (msg, font) {
    s.font = g.font = "800 82px " + that.allFonts[ font ];
    s.fillStyle = "#fff";
    s.fillRect(0, 0, width, height);

    s.fillStyle = "#222";
    s.fillText(msg, textOffsetLeft, textAscent);

    // Pull reference
    var imageData = s.getImageData(0, 0, width, height);
    pixels = imageData.data;

  };

  // Called once per frame, updates the animation.
  var render = function () {
    that.framesRendered ++;

    s.font = g.font = "800 82px " + that.allFonts[ that.font ];

    g.clearRect(0, 0, width, height);

    if (_this.displayOutline) {
      g.globalCompositeOperation = "source-over";
      g.strokeStyle = "#000";
      g.lineWidth = .5;
      g.strokeText(message, textOffsetLeft, textAscent);
    }

    g.globalCompositeOperation = "darker";

    for (var i = 0; i < particles.length; i++) {
      g.fillStyle = colors[i % colors.length];
      particles[i].render();
    }

  };

  // Returns x, y coordinates for a given index in the pixel array.
  var getPosition = function (i) {
    return {
      x: (i - (width * 4) * Math.floor(i / (width * 4))) / 4,
      y: Math.floor(i / (width * 4))
    };
  };

  // Returns a color for a given pixel in the pixel array.
  var getColor = function (x, y) {
    var base = (Math.floor(y) * width + Math.floor(x)) * 4;
    var c = {
      r: pixels[base + 0],
      g: pixels[base + 1],
      b: pixels[base + 2],
      a: pixels[base + 3]
    };

    return "rgb(" + c.r + "," + c.g + "," + c.b + ")";
  };

  // This calls the setter we've defined above, so it also calls
  // the createBitmap function.
  this.message = message;

  var loop = function() {

    requestAnimationFrame(loop);

    // Don't render if we don't see it.
    // Would be cleaner if I dynamically acquired the top of the canvas.
    if (document.body.scrollTop < height + 20) {
      render();
    }

  }

  // This calls the render function every 30 milliseconds.
  loop();

  // This class is responsible for drawing and moving those little
  // colored dots.
  function Particle(x, y, c) {

    // Position
    this.x = x;
    this.y = y;

    // Size of particle
    this.r = 0;

    // This velocity is used by the explode function.
    this.vx = 0;
    this.vy = 0;

    // Called every frame
    this.render = function () {

      // What color is the pixel we're sitting on top of?
      var c = getColor(this.x, this.y);

      // Where should we move?
      var angle = noise(this.x / noiseScale, this.y / noiseScale) * _this.noiseStrength;

      // Are we within the boundaries of the image?
      var onScreen = this.x > 0 && this.x < width &&
          this.y > 0 && this.y < height;

      var isBlack = c != "rgb(255,255,255)" && onScreen;

      // If we're on top of a black pixel, grow.
      // If not, shrink.
      if (isBlack) {
        this.r += _this.growthSpeed;
      } else {
        this.r -= _this.growthSpeed;
      }

      // This velocity is used by the explode function.
      this.vx *= 0.5;
      this.vy *= 0.5;

      // Change our position based on the flow field and our
      // explode velocity.
      this.x += Math.cos(angle) * _this.speed + this.vx;
      this.y += -Math.sin(angle) * _this.speed + this.vy;

      this.r = DAT.GUI.constrain(this.r, 0, _this.maxSize);

      // If we're tiny, keep moving around until we find a black
      // pixel.
      if (this.r <= 0) {
        this.x = Math.random() * width;
        this.y = Math.random() * height;
        return; // Don't draw!
      }

      // Draw the circle.
      g.beginPath();
      g.arc(this.x, this.y, this.r, 0, Math.PI * 2, false);
      g.fill();

    }

  }

}
